# Day 1: Trebuchet?!
Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:

```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet

In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [15]:
# read file line by line
input = open('inputs/d1input.txt', 'r')
lines = input.readlines()
cum_sum = 0

for i in range(len(lines)):
    # track first and last
    first_num = None
    last_num = None

    # iterate thru line
    for j in range(len(lines[i])):
        # get first number if char is numeric and not set
        if lines[i][j].isnumeric():
            if first_num is None:
                first_num = lines[i][j]
            last_num = lines[i][j]
    
    # parse final number and add to cumsum
    cum_sum+= int(first_num + last_num)

print(cum_sum)

54968


## Result : 54968

# Day 1, Part 2

Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:
```
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
```
In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding these together produces 281.

What is the sum of all of the calibration values?

### Smart Trie Solution (Incomplete)

I did day 1 late so I just wanted to get soemthing done, will come back and complete this though (or maybe not)

In [44]:
# create a trie(?) to store the names of numbers one through nine
class Node(object):
    def __init__(self, char, isEnd, num = None):
        self.char = char
        self.num = num
        self.isEnd = isEnd
        self.children = []

    def add_child(self, obj):
        self.children.append(obj)

numList = [
    ("one",1),
    ("two", 2),
    ("three", 3),
    ("four", 4),
    ("five", 5),
    ("six", 6),
    ("seven", 7),
    ("eight", 8),
    ("nine", 9),
]

# create a trie for the numbers
numtrie = Node(None, False)

def addWord(word, root, num=None):
    #check if root contains the first letter of the word in its children
    for child in root.children:
        if child.char == word[0]:
            # if it does, check if next letter is out of bounds
            if len(word) == 1:
                # if it is, set the node to be the end of a word
                child.num = num
                child.isEnd = True
                return
            # if it isn't, add the rest of the word to the child
            addWord(word[1:], child)
            return
        
    if len(word) > 1:
        # if it doesn't, create a new node with the first letter of the word
        new_node = Node(word[0], False)
        # add the rest of the word to the new node unless it is out of bounds
        addWord(word[1:], new_node)
        # add the new node to the root's children
        root.add_child(new_node)
        return
    else:
        # if it is, set the node to be the end of a word
        root.add_child(Node(word[0], True, num))
        return
  
# add all the numbers to the trie
for num in numList:
    addWord(num[0], numtrie, num[1])

# print trie
def printTrie(root):
    print(root.char, root.num)
    for child in root.children:
        printTrie(child)

# printTrie(numtrie)

In [39]:
# strategy- create chains of numbers that invalidate or complete a number alongside actual numbers from before


for i in range(len(lines)):
    # track first and last
    first_num = None
    last_num = None

    # iterate thru line
    chains = []
    for j in range(len(lines[i])):
        # get first number if char is numeric and not set
        if lines[i][j].isnumeric():
            if first_num is None:
                first_num = lines[i][j]
            last_num = lines[i][j]
        else:

            # then, add and validate/invalidate all existing chains
            for chain in chains:
                # get last char in chain

SyntaxError: incomplete input (331004671.py, line 22)

### Bruteforce Solution (Complete)

In [42]:
new_cum_sum = 0
input = open('inputs/d1input.txt', 'r')
lines = input.readlines()
for i in range(len(lines)):
    first_num = None
    last_num = None

    end = False
    j = 0
    # print(lines[i])
    while(not end):
        if lines[i][j].isnumeric():
            if first_num is None:
                first_num = lines[i][j]
            last_num = lines[i][j]
        else:
            # alpha, just check if it's one of the words
            for number in numList:
                # print(f"check {number[0]} at {lines[i][j]} in {lines[i]}")
                failed = False
                for charidx in range(len(number[0])):
                    
                    if j+charidx >= len(lines[i]) or number[0][charidx] != lines[i][j+charidx]:
                        # print("failed")
                        failed = True
                        break
                if not failed:
                    # print("Go ", number[0])
                    # j += len(number[0])-1
                    if(first_num is None):
                        first_num = str(number[1])
                    last_num = str(number[1])
                    break
        j+=1
        if(j >= len(lines[i])):
            new_cum_sum += int(first_num + last_num)
            end = True

print(new_cum_sum)

54094


### Answer: 54094